File: src/web-ui/src/infrastructure/api/service-api/KnowledgeAPI.ts
import { api } from './ApiClient';
import { createTauriCommandError } from '../errors/TauriCommandError';

export type KnowledgeItemType = 'file' | 'url' | 'note' | 'directory';

export type KnowledgeItemStatus =
  | 'idle'
  | 'preparing'
  | 'processing'
  | 'reading'
  | 'embedding'
  | 'completed'
  | 'failed'
  | 'deleting';

export type KnowledgeBaseStatus = 'completed' | 'failed';

export interface KnowledgeBase {
  id: string;
  name: string;
  embeddingModelId?: string | null;
  dimensions?: number | null;
  rerankModelId?: string | null;
  status: KnowledgeBaseStatus;
  error?: string | null;
  chunkSize: number;
  chunkOverlap: number;
  chunkStrategy: string;
  chunkSeparator: string;
  threshold?: number | null;
  documentCount?: number | null;
  createdAt: string;
  updatedAt: string;
}

export interface KnowledgeItem {
  id: string;
  baseId: string;
  groupId?: string | null;
  type: KnowledgeItemType;
  data: Record<string, unknown>;
  status: KnowledgeItemStatus;
  error?: string | null;
  createdAt: string;
  updatedAt: string;
}

export type KnowledgeAddItemInput =
  | { type: 'file'; groupId?: string | null; data: { source: string; path: string } }
  | { type: 'url'; groupId?: string | null; data: { source: string; url: string } }
  | { type: 'note'; groupId?: string | null; data: { source: string; content: string } }
  | { type: 'directory'; groupId?: string | null; data: { source: string } };

export interface KnowledgeChunkMetadata {
  itemId: string;
  itemType: KnowledgeItemType;
  source: string;
  chunkIndex: number;
  tokenCount: number;
}

export interface KnowledgeSearchResult {
  pageContent: string;
  score: number;
  scoreKind: string;
  rank: number;
  metadata: KnowledgeChunkMetadata;
  itemId: string;
  chunkId: string;
}

export interface KnowledgeItemChunk {
  id: string;
  itemId: string;
  content: string;
  metadata: KnowledgeChunkMetadata;
}

export interface CreateKnowledgeBaseRequest {
  name: string;
  embeddingModelId?: string | null;
  dimensions?: number | null;
  rerankModelId?: string | null;
  chunkSize?: number;
  chunkOverlap?: number;
  chunkStrategy?: string;
  chunkSeparator?: string;
  threshold?: number;
  documentCount?: number;
}

export class KnowledgeAPI {
  async createBase(request: CreateKnowledgeBaseRequest): Promise<KnowledgeBase> {
    try {
      return await api.invoke<KnowledgeBase>('knowledge_create_base', { request });
    } catch (error) {
      throw createTauriCommandError('knowledge_create_base', error, request);
    }
  }

  async listBases(): Promise<KnowledgeBase[]> {
    try {
      return await api.invoke<KnowledgeBase[]>('knowledge_list_bases');
    } catch (error) {
      throw createTauriCommandError('knowledge_list_bases', error);
    }
  }

  async getBase(baseId: string): Promise<KnowledgeBase> {
    try {
      return await api.invoke<KnowledgeBase>('knowledge_get_base', { request: { baseId } });
    } catch (error) {
      throw createTauriCommandError('knowledge_get_base', error, { baseId });
    }
  }

  async deleteBase(baseId: string): Promise<boolean> {
    try {
      return await api.invoke<boolean>('knowledge_delete_base', { request: { baseId } });
    } catch (error) {
      throw createTauriCommandError('knowledge_delete_base', error, { baseId });
    }
  }

  async addItems(baseId: string, items: KnowledgeAddItemInput[]): Promise<KnowledgeItem[]> {
    try {
      return await api.invoke<KnowledgeItem[]>('knowledge_add_items', {
        request: { baseId, items },
      });
    } catch (error) {
      throw createTauriCommandError('knowledge_add_items', error, { baseId, items });
    }
  }

  async listItems(baseId: string): Promise<KnowledgeItem[]> {
    try {
      return await api.invoke<KnowledgeItem[]>('knowledge_list_items', { request: { baseId } });
    } catch (error) {
      throw createTauriCommandError('knowledge_list_items', error, { baseId });
    }
  }

  async deleteItems(baseId: string, itemIds: string[]): Promise<boolean> {
    try {
      return await api.invoke<boolean>('knowledge_delete_items', {
        request: { baseId, itemIds },
      });
    } catch (error) {
      throw createTauriCommandError('knowledge_delete_items', error, { baseId, itemIds });
    }
  }

  async reindexItems(baseId: string, itemIds: string[]): Promise<boolean> {
    try {
      return await api.invoke<boolean>('knowledge_reindex_items', {
        request: { baseId, itemIds },
      });
    } catch (error) {
      throw createTauriCommandError('knowledge_reindex_items', error, { baseId, itemIds });
    }
  }

  async search(baseId: string, query: string): Promise<KnowledgeSearchResult[]> {
    try {
      return await api.invoke<KnowledgeSearchResult[]>('knowledge_search', {
        request: { baseId, query },
      });
    } catch (error) {
      throw createTauriCommandError('knowledge_search', error, { baseId, query });
    }
  }

  async listItemChunks(baseId: string, itemId: string): Promise<KnowledgeItemChunk[]> {
    try {
      return await api.invoke<KnowledgeItemChunk[]>('knowledge_list_item_chunks', {
        request: { baseId, itemId },
      });
    } catch (error) {
      throw createTauriCommandError('knowledge_list_item_chunks', error, { baseId, itemId });
    }
  }

  async getFilePath(itemId: string): Promise<string> {
    try {
      return await api.invoke<string>('knowledge_get_file_path', { request: { itemId } });
    } catch (error) {
      throw createTauriCommandError('knowledge_get_file_path', error, { itemId });
    }
  }
}

export const knowledgeAPI = new KnowledgeAPI();

File: src/apps/desktop/src/api/knowledge_api.rs
//! Knowledge base Tauri commands.
//!
//! Thin adapters over `bitfun_core::service::knowledge::KnowledgeService`. The
//! service owns persistence and retrieval; these commands only translate the
//! structured `request` payloads and surface errors as strings.

use bitfun_core::service::knowledge::{
    get_global_knowledge_service, CreateKnowledgeBaseRequest, KnowledgeAddItemInput, KnowledgeBase,
    KnowledgeItem, KnowledgeItemChunk, KnowledgeSearchResult, KnowledgeService,
};
use log::debug;
use serde::Deserialize;
use std::sync::Arc;

fn knowledge_service() -> Result<Arc<KnowledgeService>, String> {
    get_global_knowledge_service().ok_or_else(|| "Knowledge service is not initialized".to_string())
}

fn map_error(action: &str, error: impl std::fmt::Display) -> String {
    format!("{action} failed: {error}")
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateKnowledgeBaseCommandRequest {
    pub name: String,
    #[serde(default)]
    pub embedding_model_id: Option<String>,
    #[serde(default)]
    pub dimensions: Option<i64>,
    #[serde(default)]
    pub rerank_model_id: Option<String>,
    #[serde(default)]
    pub chunk_size: Option<i64>,
    #[serde(default)]
    pub chunk_overlap: Option<i64>,
    #[serde(default)]
    pub chunk_strategy: Option<String>,
    #[serde(default)]
    pub chunk_separator: Option<String>,
    #[serde(default)]
    pub threshold: Option<f64>,
    #[serde(default)]
    pub document_count: Option<i64>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BaseIdRequest {
    pub base_id: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddItemsRequest {
    pub base_id: String,
    pub items: Vec<KnowledgeAddItemInput>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ItemIdsRequest {
    pub base_id: String,
    pub item_ids: Vec<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchRequest {
    pub base_id: String,
    pub query: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ItemIdRequest {
    pub item_id: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListItemChunksRequest {
    pub base_id: String,
    pub item_id: String,
}

#[tauri::command]
pub async fn knowledge_create_base(
    request: CreateKnowledgeBaseCommandRequest,
) -> Result<KnowledgeBase, String> {
    debug!("Creating knowledge base: name={}", request.name);
    let service = knowledge_service()?;
    service
        .create_base(CreateKnowledgeBaseRequest {
            name: request.name,
            embedding_model_id: request.embedding_model_id,
            dimensions: request.dimensions,
            rerank_model_id: request.rerank_model_id,
            chunk_size: request.chunk_size,
            chunk_overlap: request.chunk_overlap,
            chunk_strategy: request.chunk_strategy,
            chunk_separator: request.chunk_separator,
            threshold: request.threshold,
            document_count: request.document_count,
        })
        .map_err(|error| map_error("Create knowledge base", error))
}

#[tauri::command]
pub async fn knowledge_list_bases() -> Result<Vec<KnowledgeBase>, String> {
    let service = knowledge_service()?;
    service
        .list_bases()
        .map_err(|error| map_error("List knowledge bases", error))
}

#[tauri::command]
pub async fn knowledge_get_base(request: BaseIdRequest) -> Result<KnowledgeBase, String> {
    let service = knowledge_service()?;
    service
        .get_base(&request.base_id)
        .map_err(|error| map_error("Get knowledge base", error))
}

#[tauri::command]
pub async fn knowledge_delete_base(request: BaseIdRequest) -> Result<bool, String> {
    debug!("Deleting knowledge base: base_id={}", request.base_id);
    let service = knowledge_service()?;
    service
        .delete_base(&request.base_id)
        .map(|_| true)
        .map_err(|error| map_error("Delete knowledge base", error))
}

#[tauri::command]
pub async fn knowledge_add_items(request: AddItemsRequest) -> Result<Vec<KnowledgeItem>, String> {
    debug!(
        "Adding knowledge items: base_id={}, count={}",
        request.base_id,
        request.items.len()
    );
    let service = knowledge_service()?;
    service
        .add_items(&request.base_id, request.items)
        .map_err(|error| map_error("Add knowledge items", error))
}

#[tauri::command]
pub async fn knowledge_list_items(request: BaseIdRequest) -> Result<Vec<KnowledgeItem>, String> {
    let service = knowledge_service()?;
    service
        .list_items(&request.base_id)
        .map_err(|error| map_error("List knowledge items", error))
}

#[tauri::command]
pub async fn knowledge_delete_items(request: ItemIdsRequest) -> Result<bool, String> {
    let service = knowledge_service()?;
    service
        .delete_items(&request.base_id, request.item_ids)
        .map(|_| true)
        .map_err(|error| map_error("Delete knowledge items", error))
}

#[tauri::command]
pub async fn knowledge_reindex_items(request: ItemIdsRequest) -> Result<bool, String> {
    let service = knowledge_service()?;
    service
        .reindex_items(&request.base_id, request.item_ids)
        .map(|_| true)
        .map_err(|error| map_error("Reindex knowledge items", error))
}

#[tauri::command]
pub async fn knowledge_search(
    request: SearchRequest,
) -> Result<Vec<KnowledgeSearchResult>, String> {
    let service = knowledge_service()?;
    service
        .search(&request.base_id, &request.query)
        .await
        .map_err(|error| map_error("Search knowledge base", error))
}

#[tauri::command]
pub async fn knowledge_list_item_chunks(
    request: ListItemChunksRequest,
) -> Result<Vec<KnowledgeItemChunk>, String> {
    let service = knowledge_service()?;
    service
        .list_item_chunks(&request.base_id, &request.item_id)
        .map_err(|error| map_error("List knowledge item chunks", error))
}

#[tauri::command]
pub async fn knowledge_get_file_path(request: ItemIdRequest) -> Result<String, String> {
    let service = knowledge_service()?;
    service
        .get_file_path(&request.item_id)
        .map(|path| path.to_string_lossy().to_string())
        .map_err(|error| map_error("Get knowledge file path", error))
}
